{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffutils\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This can take a long time, so only run this once (remove the #)\n",
    "\n",
    "#db = gffutils.create_db('chm13.draft_v2.0.gene_annotation.gff3', dbfn='chm13db.db', force=True, keep_order=True, \n",
    "#merge_strategy='create_unique', sort_attribute_values=True) \n",
    "\n",
    "#replace gff3 with your gff3 file, dbfn can be whichever name you want to store the transposed file as\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gffutils.interface.FeatureDB at 0x7f59db00e790>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db = gffutils.FeatureDB('chm13db.db', keep_order=True)\n",
    "db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "slct = db.region(region=('chr1', 207320867, 207320897), completely_within=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for gene in slct:\n",
    "    records.append({\n",
    "        \"seqid\": gene.seqid,\n",
    "        \"source\": gene.source,\n",
    "        \"featuretype\": gene.featuretype,\n",
    "        \"start\": gene.start,\n",
    "        \"end\": gene.end,\n",
    "        \"strand\": gene.strand,\n",
    "        \"score\": gene.score,\n",
    "        \"attributes\": gene.attributes  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>source_transcript</th>\n",
       "      <th>source_transcript_name</th>\n",
       "      <th>source_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>collapsed_gene_ids</th>\n",
       "      <th>collapsed_gene_names</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>exon_anotation_support</th>\n",
       "      <th>novel_5p_cap</th>\n",
       "      <th>novel_poly_a</th>\n",
       "      <th>intron_rna_support</th>\n",
       "      <th>exon_rna_support</th>\n",
       "      <th>rna_support</th>\n",
       "      <th>reference_support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207268990</td>\n",
       "      <td>207491104</td>\n",
       "      <td>-</td>\n",
       "      <td>9980</td>\n",
       "      <td>[ENST00000367033.4]</td>\n",
       "      <td>[PLXNA2-201]</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>gene</td>\n",
       "      <td>207268990</td>\n",
       "      <td>207491209</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207273418</td>\n",
       "      <td>207491209</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207273418</td>\n",
       "      <td>207491209</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>207307614</td>\n",
       "      <td>207326032</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>[ENST00000367033.4]</td>\n",
       "      <td>[PLXNA2-201]</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[True]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>207307614</td>\n",
       "      <td>207326032</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[True]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>207307614</td>\n",
       "      <td>207326032</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000076356.7]</td>\n",
       "      <td>...</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[N/A]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[True]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows Ã— 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid source featuretype      start        end strand score  \\\n",
       "0  chr1    CAT  transcript  207268990  207491104      -  9980   \n",
       "1  chr1    CAT        gene  207268990  207491209      -     .   \n",
       "2  chr1    CAT  transcript  207273418  207491209      -     .   \n",
       "3  chr1    CAT  transcript  207273418  207491209      -     .   \n",
       "4  chr1    CAT      intron  207307614  207326032      -     .   \n",
       "5  chr1    CAT      intron  207307614  207326032      -     .   \n",
       "6  chr1    CAT      intron  207307614  207326032      -     .   \n",
       "\n",
       "     source_transcript source_transcript_name          source_gene  ...  \\\n",
       "0  [ENST00000367033.4]           [PLXNA2-201]  [ENSG00000076356.7]  ...   \n",
       "1                [N/A]                    NaN  [ENSG00000076356.7]  ...   \n",
       "2                [N/A]                    NaN  [ENSG00000076356.7]  ...   \n",
       "3                [N/A]                    NaN  [ENSG00000076356.7]  ...   \n",
       "4  [ENST00000367033.4]           [PLXNA2-201]  [ENSG00000076356.7]  ...   \n",
       "5                [N/A]                    NaN  [ENSG00000076356.7]  ...   \n",
       "6                [N/A]                    NaN  [ENSG00000076356.7]  ...   \n",
       "\n",
       "  collapsed_gene_ids collapsed_gene_names extra_paralog  \\\n",
       "0              [N/A]                [N/A]       [False]   \n",
       "1              [N/A]                [N/A]       [False]   \n",
       "2              [N/A]                [N/A]       [False]   \n",
       "3              [N/A]                [N/A]       [False]   \n",
       "4              [N/A]                [N/A]       [False]   \n",
       "5              [N/A]                [N/A]       [False]   \n",
       "6              [N/A]                [N/A]       [False]   \n",
       "\n",
       "  exon_anotation_support novel_5p_cap novel_poly_a  \\\n",
       "0                    NaN          NaN          NaN   \n",
       "1                  [N/A]          NaN          NaN   \n",
       "2                    NaN       [True]       [True]   \n",
       "3                    NaN       [True]       [True]   \n",
       "4                    NaN          NaN          NaN   \n",
       "5                    NaN       [True]       [True]   \n",
       "6                    NaN       [True]       [True]   \n",
       "\n",
       "                                  intron_rna_support  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "3  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "4                                                NaN   \n",
       "5  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "6  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "\n",
       "                                    exon_rna_support rna_support  \\\n",
       "0                                                NaN         NaN   \n",
       "1                                                NaN         NaN   \n",
       "2  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...         NaN   \n",
       "3  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...         NaN   \n",
       "4                                                NaN       [N/A]   \n",
       "5  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...      [True]   \n",
       "6  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...      [True]   \n",
       "\n",
       "  reference_support  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4            [True]  \n",
       "5            [True]  \n",
       "6            [True]  \n",
       "\n",
       "[7 rows x 52 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.DataFrame(records)\n",
    "attributes_df = df[\"attributes\"].apply(pd.Series)\n",
    "df = pd.concat([df.drop(columns=[\"attributes\"]), attributes_df], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>source_transcript</th>\n",
       "      <th>source_transcript_name</th>\n",
       "      <th>source_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>collapsed_gene_ids</th>\n",
       "      <th>collapsed_gene_names</th>\n",
       "      <th>extra_paralog</th>\n",
       "      <th>exon_anotation_support</th>\n",
       "      <th>novel_5p_cap</th>\n",
       "      <th>novel_poly_a</th>\n",
       "      <th>intron_rna_support</th>\n",
       "      <th>exon_rna_support</th>\n",
       "      <th>rna_support</th>\n",
       "      <th>reference_support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207268990</td>\n",
       "      <td>207491104</td>\n",
       "      <td>-</td>\n",
       "      <td>9980</td>\n",
       "      <td>ENST00000367033.4</td>\n",
       "      <td>PLXNA2-201</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>gene</td>\n",
       "      <td>207268990</td>\n",
       "      <td>207491209</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207273418</td>\n",
       "      <td>207491209</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207273418</td>\n",
       "      <td>207491209</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>207307614</td>\n",
       "      <td>207326032</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000367033.4</td>\n",
       "      <td>PLXNA2-201</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N/A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>207307614</td>\n",
       "      <td>207326032</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CAT</td>\n",
       "      <td>intron</td>\n",
       "      <td>207307614</td>\n",
       "      <td>207326032</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000076356.7</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows Ã— 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid source featuretype      start        end strand score  \\\n",
       "0  chr1    CAT  transcript  207268990  207491104      -  9980   \n",
       "1  chr1    CAT        gene  207268990  207491209      -     .   \n",
       "2  chr1    CAT  transcript  207273418  207491209      -     .   \n",
       "3  chr1    CAT  transcript  207273418  207491209      -     .   \n",
       "4  chr1    CAT      intron  207307614  207326032      -     .   \n",
       "5  chr1    CAT      intron  207307614  207326032      -     .   \n",
       "6  chr1    CAT      intron  207307614  207326032      -     .   \n",
       "\n",
       "   source_transcript source_transcript_name        source_gene  ...  \\\n",
       "0  ENST00000367033.4             PLXNA2-201  ENSG00000076356.7  ...   \n",
       "1                N/A                    NaN  ENSG00000076356.7  ...   \n",
       "2                N/A                    NaN  ENSG00000076356.7  ...   \n",
       "3                N/A                    NaN  ENSG00000076356.7  ...   \n",
       "4  ENST00000367033.4             PLXNA2-201  ENSG00000076356.7  ...   \n",
       "5                N/A                    NaN  ENSG00000076356.7  ...   \n",
       "6                N/A                    NaN  ENSG00000076356.7  ...   \n",
       "\n",
       "  collapsed_gene_ids collapsed_gene_names extra_paralog  \\\n",
       "0                N/A                  N/A         False   \n",
       "1                N/A                  N/A         False   \n",
       "2                N/A                  N/A         False   \n",
       "3                N/A                  N/A         False   \n",
       "4                N/A                  N/A         False   \n",
       "5                N/A                  N/A         False   \n",
       "6                N/A                  N/A         False   \n",
       "\n",
       "  exon_anotation_support novel_5p_cap novel_poly_a  \\\n",
       "0                    NaN          NaN          NaN   \n",
       "1                    N/A          NaN          NaN   \n",
       "2                    NaN         True         True   \n",
       "3                    NaN         True         True   \n",
       "4                    NaN          NaN          NaN   \n",
       "5                    NaN         True         True   \n",
       "6                    NaN         True         True   \n",
       "\n",
       "                                  intron_rna_support  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "3  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "4                                                NaN   \n",
       "5  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "6  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   \n",
       "\n",
       "                                    exon_rna_support rna_support  \\\n",
       "0                                                NaN         NaN   \n",
       "1                                                NaN         NaN   \n",
       "2  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...         NaN   \n",
       "3  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...         NaN   \n",
       "4                                                NaN         N/A   \n",
       "5  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...        True   \n",
       "6  [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...        True   \n",
       "\n",
       "  reference_support  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4              True  \n",
       "5              True  \n",
       "6              True  \n",
       "\n",
       "[7 rows x 52 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for col in df.columns:\n",
    "    df[col] = df[col].apply(lambda x: x[0] if isinstance(x, list) and len(x) == 1 else x)\n",
    "df.dropna(how='all', axis=1, inplace=True)\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
