{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotation Selection of the hg38 genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Before you begin, please make sure you have the proper files downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download link for gencode annotation file (gff3): https://www.gencodegenes.org/human/release_38.html\n",
    "##Make sure to download the \"comprehensive gene annotation file (CHR)\" in the gff3 format\n",
    "\n",
    "#Download link for hg38 archetype motif file: https://www.vierstra.org/resources/motif_clustering\n",
    "##Download the hg38 archetype scan file alongside its tabix file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffutils\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### If this is your first time importing the gff3 database, run this command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This can take a long time, so only run this once (remove the #)\n",
    "\n",
    "#db = gffutils.create_db('gencode.v38.annotation.gff3', dbfn='gff3db.db', force=True, keep_order=True, \n",
    "#merge_strategy='create_unique', sort_attribute_values=True) \n",
    "\n",
    "#replace gff3 with your gff3 file, dbfn can be whichever name you want to store the transposed file as\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Feature the selected database "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB('gff3db.db', keep_order=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Select the desired range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr = 'chr1' #Replace with wanted chromosome location\n",
    "start = 207320867 #Replace with start coordinate\n",
    "stop = 207320897 #Replace with end coordinate\n",
    "slct = db.region(region=(chr, start, stop), completely_within=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Iterate through the region and transpose it to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for gene in slct:\n",
    "    records.append({\n",
    "        \"seqid\": gene.seqid,\n",
    "        \"source\": gene.source,\n",
    "        \"featuretype\": gene.featuretype,\n",
    "        \"start\": gene.start,\n",
    "        \"end\": gene.end,\n",
    "        \"strand\": gene.strand,\n",
    "        \"score\": gene.score,\n",
    "        \"attributes\": gene.attributes  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Next, lets integrate the hg38 regulatory elements into this database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = 'hg38.archetype_motifs.v1.0.bed.gz' #Replace with the RE file\n",
    "tabix = pysam.TabixFile(input_file)\n",
    "source_name = 'atlas' #replace with reference cell type\n",
    "#It may say that the index file is older than the data file, you can ignore this for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create dictionary using the input selection\n",
    "for row in tabix.fetch(chr,start,stop):\n",
    "    assets = row.split()\n",
    "    records.append({\n",
    "        \"seqid\": assets.pop(0),\n",
    "        \"source\": source_name,\n",
    "        \"start\": int(assets.pop(0)),\n",
    "        \"end\": int(assets.pop(0)),\n",
    "        \"featuretype\": 'RE',\n",
    "        \"strand\": assets.pop(2),\n",
    "        \"attributes\": ({\"ID\": assets.pop(2), \"matchscore\": float(assets.pop(1)), \"cluster\":assets.pop(0)})  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Expand out the attributes and format the dictionary as a database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320851</td>\n",
       "      <td>207320868</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>SALL4_HUMAN.H11MO.0.B</td>\n",
       "      <td>9.9542</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320852</td>\n",
       "      <td>207320869</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>EGR2_HUMAN.H11MO.0.A</td>\n",
       "      <td>11.5015</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320853</td>\n",
       "      <td>207320873</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>KLF15_HUMAN.H11MO.0.A</td>\n",
       "      <td>11.5516</td>\n",
       "      <td>GC-tract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320854</td>\n",
       "      <td>207320872</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>TBX3_MOUSE.H11MO.0.B</td>\n",
       "      <td>7.7390</td>\n",
       "      <td>TBX/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320855</td>\n",
       "      <td>207320872</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>KLF1_MOUSE.H11MO.0.A</td>\n",
       "      <td>10.1096</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320893</td>\n",
       "      <td>207320910</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>EGR2_HUMAN.H11MO.0.A</td>\n",
       "      <td>7.8496</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320894</td>\n",
       "      <td>207320907</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>ZSCAN4_C2H2_1</td>\n",
       "      <td>3.8977</td>\n",
       "      <td>ZSCAN4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320895</td>\n",
       "      <td>207320906</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>KLF9_MA1107.1</td>\n",
       "      <td>8.4085</td>\n",
       "      <td>KLF/SP/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320895</td>\n",
       "      <td>207320912</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>EGR2_HUMAN.H11MO.0.A</td>\n",
       "      <td>7.8496</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320896</td>\n",
       "      <td>207320909</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>ZSCAN4_C2H2_1</td>\n",
       "      <td>3.8977</td>\n",
       "      <td>ZSCAN4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source      start        end featuretype strand  \\\n",
       "0   chr1  atlas  207320851  207320868          RE      -   \n",
       "1   chr1  atlas  207320852  207320869          RE      -   \n",
       "2   chr1  atlas  207320853  207320873          RE      -   \n",
       "3   chr1  atlas  207320854  207320872          RE      +   \n",
       "4   chr1  atlas  207320855  207320872          RE      -   \n",
       "..   ...    ...        ...        ...         ...    ...   \n",
       "62  chr1  atlas  207320893  207320910          RE      -   \n",
       "63  chr1  atlas  207320894  207320907          RE      +   \n",
       "64  chr1  atlas  207320895  207320906          RE      +   \n",
       "65  chr1  atlas  207320895  207320912          RE      -   \n",
       "66  chr1  atlas  207320896  207320909          RE      +   \n",
       "\n",
       "                       ID  matchscore   cluster  \n",
       "0   SALL4_HUMAN.H11MO.0.B      9.9542  KLF/SP/2  \n",
       "1    EGR2_HUMAN.H11MO.0.A     11.5015  KLF/SP/2  \n",
       "2   KLF15_HUMAN.H11MO.0.A     11.5516  GC-tract  \n",
       "3    TBX3_MOUSE.H11MO.0.B      7.7390     TBX/3  \n",
       "4    KLF1_MOUSE.H11MO.0.A     10.1096  KLF/SP/2  \n",
       "..                    ...         ...       ...  \n",
       "62   EGR2_HUMAN.H11MO.0.A      7.8496  KLF/SP/2  \n",
       "63          ZSCAN4_C2H2_1      3.8977    ZSCAN4  \n",
       "64          KLF9_MA1107.1      8.4085  KLF/SP/1  \n",
       "65   EGR2_HUMAN.H11MO.0.A      7.8496  KLF/SP/2  \n",
       "66          ZSCAN4_C2H2_1      3.8977    ZSCAN4  \n",
       "\n",
       "[67 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.DataFrame(records)\n",
    "attributes_df = df[\"attributes\"].apply(pd.Series)\n",
    "df = pd.concat([df.drop(columns=[\"attributes\"]), attributes_df], axis=1)\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Expand out the single list attributes as strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320851</td>\n",
       "      <td>207320868</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>SALL4_HUMAN.H11MO.0.B</td>\n",
       "      <td>9.9542</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320852</td>\n",
       "      <td>207320869</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>EGR2_HUMAN.H11MO.0.A</td>\n",
       "      <td>11.5015</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320853</td>\n",
       "      <td>207320873</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>KLF15_HUMAN.H11MO.0.A</td>\n",
       "      <td>11.5516</td>\n",
       "      <td>GC-tract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320854</td>\n",
       "      <td>207320872</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>TBX3_MOUSE.H11MO.0.B</td>\n",
       "      <td>7.7390</td>\n",
       "      <td>TBX/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320855</td>\n",
       "      <td>207320872</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>KLF1_MOUSE.H11MO.0.A</td>\n",
       "      <td>10.1096</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320893</td>\n",
       "      <td>207320910</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>EGR2_HUMAN.H11MO.0.A</td>\n",
       "      <td>7.8496</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320894</td>\n",
       "      <td>207320907</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>ZSCAN4_C2H2_1</td>\n",
       "      <td>3.8977</td>\n",
       "      <td>ZSCAN4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320895</td>\n",
       "      <td>207320906</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>KLF9_MA1107.1</td>\n",
       "      <td>8.4085</td>\n",
       "      <td>KLF/SP/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320895</td>\n",
       "      <td>207320912</td>\n",
       "      <td>RE</td>\n",
       "      <td>-</td>\n",
       "      <td>EGR2_HUMAN.H11MO.0.A</td>\n",
       "      <td>7.8496</td>\n",
       "      <td>KLF/SP/2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320896</td>\n",
       "      <td>207320909</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>ZSCAN4_C2H2_1</td>\n",
       "      <td>3.8977</td>\n",
       "      <td>ZSCAN4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source      start        end featuretype strand  \\\n",
       "0   chr1  atlas  207320851  207320868          RE      -   \n",
       "1   chr1  atlas  207320852  207320869          RE      -   \n",
       "2   chr1  atlas  207320853  207320873          RE      -   \n",
       "3   chr1  atlas  207320854  207320872          RE      +   \n",
       "4   chr1  atlas  207320855  207320872          RE      -   \n",
       "..   ...    ...        ...        ...         ...    ...   \n",
       "62  chr1  atlas  207320893  207320910          RE      -   \n",
       "63  chr1  atlas  207320894  207320907          RE      +   \n",
       "64  chr1  atlas  207320895  207320906          RE      +   \n",
       "65  chr1  atlas  207320895  207320912          RE      -   \n",
       "66  chr1  atlas  207320896  207320909          RE      +   \n",
       "\n",
       "                       ID  matchscore   cluster  \n",
       "0   SALL4_HUMAN.H11MO.0.B      9.9542  KLF/SP/2  \n",
       "1    EGR2_HUMAN.H11MO.0.A     11.5015  KLF/SP/2  \n",
       "2   KLF15_HUMAN.H11MO.0.A     11.5516  GC-tract  \n",
       "3    TBX3_MOUSE.H11MO.0.B      7.7390     TBX/3  \n",
       "4    KLF1_MOUSE.H11MO.0.A     10.1096  KLF/SP/2  \n",
       "..                    ...         ...       ...  \n",
       "62   EGR2_HUMAN.H11MO.0.A      7.8496  KLF/SP/2  \n",
       "63          ZSCAN4_C2H2_1      3.8977    ZSCAN4  \n",
       "64          KLF9_MA1107.1      8.4085  KLF/SP/1  \n",
       "65   EGR2_HUMAN.H11MO.0.A      7.8496  KLF/SP/2  \n",
       "66          ZSCAN4_C2H2_1      3.8977    ZSCAN4  \n",
       "\n",
       "[67 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for col in df.columns:\n",
    "    df[col] = df[col].apply(lambda x: x[0] if isinstance(x, list) and len(x) == 1 else x)\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Use the following code to find an RE from the human genome browser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_235/981348799.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  lst_of_id.dropna(how='all', axis=1, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320872</td>\n",
       "      <td>207320883</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>ZFX_MOUSE.H11MO.0.B</td>\n",
       "      <td>7.7766</td>\n",
       "      <td>ZFX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source      start        end featuretype strand                   ID  \\\n",
       "26  chr1  atlas  207320872  207320883          RE      +  ZFX_MOUSE.H11MO.0.B   \n",
       "\n",
       "    matchscore cluster  \n",
       "26      7.7766     ZFX  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_of_id = df[df[\"ID\"] == 'ZFX_MOUSE.H11MO.0.B'] ##Replace with the RE ID\n",
    "lst_of_id.dropna(how='all', axis=1, inplace=True)\n",
    "lst_of_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>atlas</td>\n",
       "      <td>207320872</td>\n",
       "      <td>207320883</td>\n",
       "      <td>RE</td>\n",
       "      <td>+</td>\n",
       "      <td>ZFX_MOUSE.H11MO.0.B</td>\n",
       "      <td>7.7766</td>\n",
       "      <td>ZFX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source      start        end featuretype strand                   ID  \\\n",
       "26  chr1  atlas  207320872  207320883          RE      +  ZFX_MOUSE.H11MO.0.B   \n",
       "\n",
       "    matchscore cluster  \n",
       "26      7.7766     ZFX  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_of_id[lst_of_id['matchscore'] == 7.7766] ###Replace with matchscore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
