{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Annotation Selection of the hg38 genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffutils\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### If this is your first time importing the gff3 database, run this command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This can take a long time, so only run this once (remove the #)\n",
    "\n",
    "#db = gffutils.create_db('gencode.v38.annotation.gff3', dbfn='test2.db', force=True, keep_order=True, \n",
    "#merge_strategy='keep_unique', sort_attribute_values=True) \n",
    "\n",
    "#replace gff3 with your gff3 file, dbfn can be whichever name you want to store the transposed file as\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Feature the selected database "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB('test2.db', keep_order=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Select the desired range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr = 'chr1' #Replace with wanted chromosome location\n",
    "start = 207317678 #Replace with start coordinate\n",
    "stop = 207325678 #Replace with end coordinate\n",
    "slct = db.region(region=(chr, start, stop), completely_within=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Iterate through the region and transpose it to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for gene in slct:\n",
    "    records.append({\n",
    "        \"seqid\": gene.seqid,\n",
    "        \"source\": gene.source,\n",
    "        \"featuretype\": gene.featuretype,\n",
    "        \"start\": gene.start,\n",
    "        \"end\": gene.end,\n",
    "        \"strand\": gene.strand,\n",
    "        \"score\": gene.score,\n",
    "        \"attributes\": gene.attributes  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Next, lets integrate the cell-specific regulatory elements into this database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::hts_idx_load3] The index file is older than the data file: hg38.all_motifs.v1.0.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "input_file = 'hg38.all_motifs.v1.0.bed.gz' #Replace with the RE file\n",
    "tabix = pysam.TabixFile(input_file)\n",
    "cell_type = 'K562' #replace with reference cell type\n",
    "#It may say that the index file is older than the data file, you can ignore this for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create dictionary using the input selection\n",
    "for row in tabix.fetch(chr,start,stop):\n",
    "    assets = row.split()\n",
    "    records.append({\n",
    "        \"seqid\": assets.pop(0),\n",
    "        \"source\": cell_type,\n",
    "        \"start\": int(assets.pop(0)),\n",
    "        \"end\": int(assets.pop(0)),\n",
    "        \"featuretype\": 'RE',\n",
    "        \"strand\": assets.pop(2),\n",
    "        \"attributes\": ({\"ID\": assets.pop(0), \"matchscore\": float(assets.pop(0)), \"seq\":assets.pop(0)})  # Attributes as a dictionary (will be expanded later)\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Expand out the attributes and format the dictionary as a database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>level</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207321765</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>[UTR5:ENST00000367063.6]</td>\n",
       "      <td>[ENST00000367063.6]</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>...</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[ENSP00000356030.2]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[HGNC:2665]</td>\n",
       "      <td>[basic, CCDS]</td>\n",
       "      <td>[CCDS73022.1]</td>\n",
       "      <td>[OTTHUMG00000036255.7]</td>\n",
       "      <td>[OTTHUMT00000088210.2]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207321865</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>[exon:ENST00000367063.6:1]</td>\n",
       "      <td>[ENST00000367063.6]</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>...</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[ENSP00000356030.2]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[HGNC:2665]</td>\n",
       "      <td>[basic, CCDS]</td>\n",
       "      <td>[CCDS73022.1]</td>\n",
       "      <td>[OTTHUMG00000036255.7]</td>\n",
       "      <td>[OTTHUMT00000088210.2]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207340766</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>[ENST00000367063.6]</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>...</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[ENSP00000356030.2]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[HGNC:2665]</td>\n",
       "      <td>[basic, CCDS]</td>\n",
       "      <td>[CCDS73022.1]</td>\n",
       "      <td>[OTTHUMG00000036255.7]</td>\n",
       "      <td>[OTTHUMT00000088210.2]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207386804</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[HGNC:2665]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[OTTHUMG00000036255.7]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>207321643</td>\n",
       "      <td>207321765</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>[UTR5:ENST00000391921.9]</td>\n",
       "      <td>[ENST00000391921.9]</td>\n",
       "      <td>[ENSG00000196352.16]</td>\n",
       "      <td>...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[ENSP00000375788.4]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[HGNC:2665]</td>\n",
       "      <td>[basic, exp_conf]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[OTTHUMG00000036255.7]</td>\n",
       "      <td>[OTTHUMT00000088209.3]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325670</td>\n",
       "      <td>207325680</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI2_HUMAN.H11MO.0.A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.703112</td>\n",
       "      <td>ACCAGGTGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7243</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325671</td>\n",
       "      <td>207325679</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI1_HUMAN.H11MO.0.C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.495118</td>\n",
       "      <td>CCAGGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7244</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325671</td>\n",
       "      <td>207325679</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI1_MOUSE.H11MO.0.C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.495118</td>\n",
       "      <td>CCAGGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7245</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325673</td>\n",
       "      <td>207325682</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hic1_MA0739.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.212763</td>\n",
       "      <td>AGGTGGCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7246</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325673</td>\n",
       "      <td>207325682</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hic1.mouse_C2H2_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.212763</td>\n",
       "      <td>AGGTGGCAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7247 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid  source     featuretype      start        end strand score  \\\n",
       "0     chr1  HAVANA  five_prime_UTR  207321532  207321765      +     .   \n",
       "1     chr1  HAVANA            exon  207321532  207321865      +     .   \n",
       "2     chr1  HAVANA      transcript  207321532  207340766      +     .   \n",
       "3     chr1  HAVANA            gene  207321532  207386804      +     .   \n",
       "4     chr1  HAVANA  five_prime_UTR  207321643  207321765      +     .   \n",
       "...    ...     ...             ...        ...        ...    ...   ...   \n",
       "7242  chr1    K562              RE  207325670  207325680      +   NaN   \n",
       "7243  chr1    K562              RE  207325671  207325679      +   NaN   \n",
       "7244  chr1    K562              RE  207325671  207325679      +   NaN   \n",
       "7245  chr1    K562              RE  207325673  207325682      -   NaN   \n",
       "7246  chr1    K562              RE  207325673  207325682      -   NaN   \n",
       "\n",
       "                              ID                Parent               gene_id  \\\n",
       "0       [UTR5:ENST00000367063.6]   [ENST00000367063.6]  [ENSG00000196352.16]   \n",
       "1     [exon:ENST00000367063.6:1]   [ENST00000367063.6]  [ENSG00000196352.16]   \n",
       "2            [ENST00000367063.6]  [ENSG00000196352.16]  [ENSG00000196352.16]   \n",
       "3           [ENSG00000196352.16]                   NaN  [ENSG00000196352.16]   \n",
       "4       [UTR5:ENST00000391921.9]   [ENST00000391921.9]  [ENSG00000196352.16]   \n",
       "...                          ...                   ...                   ...   \n",
       "7242       SNAI2_HUMAN.H11MO.0.A                   NaN                   NaN   \n",
       "7243       SNAI1_HUMAN.H11MO.0.C                   NaN                   NaN   \n",
       "7244       SNAI1_MOUSE.H11MO.0.C                   NaN                   NaN   \n",
       "7245               Hic1_MA0739.1                   NaN                   NaN   \n",
       "7246           Hic1.mouse_C2H2_2                   NaN                   NaN   \n",
       "\n",
       "      ... level           protein_id transcript_support_level      hgnc_id  \\\n",
       "0     ...   [2]  [ENSP00000356030.2]                      [1]  [HGNC:2665]   \n",
       "1     ...   [2]  [ENSP00000356030.2]                      [1]  [HGNC:2665]   \n",
       "2     ...   [2]  [ENSP00000356030.2]                      [1]  [HGNC:2665]   \n",
       "3     ...   [1]                  NaN                      NaN  [HGNC:2665]   \n",
       "4     ...   [1]  [ENSP00000375788.4]                      [1]  [HGNC:2665]   \n",
       "...   ...   ...                  ...                      ...          ...   \n",
       "7242  ...   NaN                  NaN                      NaN          NaN   \n",
       "7243  ...   NaN                  NaN                      NaN          NaN   \n",
       "7244  ...   NaN                  NaN                      NaN          NaN   \n",
       "7245  ...   NaN                  NaN                      NaN          NaN   \n",
       "7246  ...   NaN                  NaN                      NaN          NaN   \n",
       "\n",
       "                    tag         ccdsid             havana_gene  \\\n",
       "0         [basic, CCDS]  [CCDS73022.1]  [OTTHUMG00000036255.7]   \n",
       "1         [basic, CCDS]  [CCDS73022.1]  [OTTHUMG00000036255.7]   \n",
       "2         [basic, CCDS]  [CCDS73022.1]  [OTTHUMG00000036255.7]   \n",
       "3                   NaN            NaN  [OTTHUMG00000036255.7]   \n",
       "4     [basic, exp_conf]            NaN  [OTTHUMG00000036255.7]   \n",
       "...                 ...            ...                     ...   \n",
       "7242                NaN            NaN                     NaN   \n",
       "7243                NaN            NaN                     NaN   \n",
       "7244                NaN            NaN                     NaN   \n",
       "7245                NaN            NaN                     NaN   \n",
       "7246                NaN            NaN                     NaN   \n",
       "\n",
       "           havana_transcript matchscore         seq  \n",
       "0     [OTTHUMT00000088210.2]        NaN         NaN  \n",
       "1     [OTTHUMT00000088210.2]        NaN         NaN  \n",
       "2     [OTTHUMT00000088210.2]        NaN         NaN  \n",
       "3                        NaN        NaN         NaN  \n",
       "4     [OTTHUMT00000088209.3]        NaN         NaN  \n",
       "...                      ...        ...         ...  \n",
       "7242                     NaN   5.703112  ACCAGGTGGC  \n",
       "7243                     NaN  10.495118    CCAGGTGG  \n",
       "7244                     NaN  10.495118    CCAGGTGG  \n",
       "7245                     NaN   8.212763   AGGTGGCAT  \n",
       "7246                     NaN   8.212763   AGGTGGCAT  \n",
       "\n",
       "[7247 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.DataFrame(records)\n",
    "attributes_df = df[\"attributes\"].apply(pd.Series)\n",
    "df = pd.concat([df.drop(columns=[\"attributes\"]), attributes_df], axis=1)\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Expand out the single list attributes as strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>level</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207321765</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENST00000367063.6</td>\n",
       "      <td>ENST00000367063.6</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000356030.2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>[basic, CCDS]</td>\n",
       "      <td>CCDS73022.1</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000088210.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207321865</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000367063.6:1</td>\n",
       "      <td>ENST00000367063.6</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000356030.2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>[basic, CCDS]</td>\n",
       "      <td>CCDS73022.1</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000088210.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207340766</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000367063.6</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSP00000356030.2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>[basic, CCDS]</td>\n",
       "      <td>CCDS73022.1</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000088210.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>207321532</td>\n",
       "      <td>207386804</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>207321643</td>\n",
       "      <td>207321765</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENST00000391921.9</td>\n",
       "      <td>ENST00000391921.9</td>\n",
       "      <td>ENSG00000196352.16</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSP00000375788.4</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:2665</td>\n",
       "      <td>[basic, exp_conf]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000036255.7</td>\n",
       "      <td>OTTHUMT00000088209.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325670</td>\n",
       "      <td>207325680</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI2_HUMAN.H11MO.0.A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.703112</td>\n",
       "      <td>ACCAGGTGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7243</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325671</td>\n",
       "      <td>207325679</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI1_HUMAN.H11MO.0.C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.495118</td>\n",
       "      <td>CCAGGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7244</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325671</td>\n",
       "      <td>207325679</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI1_MOUSE.H11MO.0.C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.495118</td>\n",
       "      <td>CCAGGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7245</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325673</td>\n",
       "      <td>207325682</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hic1_MA0739.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.212763</td>\n",
       "      <td>AGGTGGCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7246</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207325673</td>\n",
       "      <td>207325682</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hic1.mouse_C2H2_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.212763</td>\n",
       "      <td>AGGTGGCAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7247 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid  source     featuretype      start        end strand score  \\\n",
       "0     chr1  HAVANA  five_prime_UTR  207321532  207321765      +     .   \n",
       "1     chr1  HAVANA            exon  207321532  207321865      +     .   \n",
       "2     chr1  HAVANA      transcript  207321532  207340766      +     .   \n",
       "3     chr1  HAVANA            gene  207321532  207386804      +     .   \n",
       "4     chr1  HAVANA  five_prime_UTR  207321643  207321765      +     .   \n",
       "...    ...     ...             ...        ...        ...    ...   ...   \n",
       "7242  chr1    K562              RE  207325670  207325680      +   NaN   \n",
       "7243  chr1    K562              RE  207325671  207325679      +   NaN   \n",
       "7244  chr1    K562              RE  207325671  207325679      +   NaN   \n",
       "7245  chr1    K562              RE  207325673  207325682      -   NaN   \n",
       "7246  chr1    K562              RE  207325673  207325682      -   NaN   \n",
       "\n",
       "                            ID              Parent             gene_id  ...  \\\n",
       "0       UTR5:ENST00000367063.6   ENST00000367063.6  ENSG00000196352.16  ...   \n",
       "1     exon:ENST00000367063.6:1   ENST00000367063.6  ENSG00000196352.16  ...   \n",
       "2            ENST00000367063.6  ENSG00000196352.16  ENSG00000196352.16  ...   \n",
       "3           ENSG00000196352.16                 NaN  ENSG00000196352.16  ...   \n",
       "4       UTR5:ENST00000391921.9   ENST00000391921.9  ENSG00000196352.16  ...   \n",
       "...                        ...                 ...                 ...  ...   \n",
       "7242     SNAI2_HUMAN.H11MO.0.A                 NaN                 NaN  ...   \n",
       "7243     SNAI1_HUMAN.H11MO.0.C                 NaN                 NaN  ...   \n",
       "7244     SNAI1_MOUSE.H11MO.0.C                 NaN                 NaN  ...   \n",
       "7245             Hic1_MA0739.1                 NaN                 NaN  ...   \n",
       "7246         Hic1.mouse_C2H2_2                 NaN                 NaN  ...   \n",
       "\n",
       "     level         protein_id transcript_support_level    hgnc_id  \\\n",
       "0        2  ENSP00000356030.2                        1  HGNC:2665   \n",
       "1        2  ENSP00000356030.2                        1  HGNC:2665   \n",
       "2        2  ENSP00000356030.2                        1  HGNC:2665   \n",
       "3        1                NaN                      NaN  HGNC:2665   \n",
       "4        1  ENSP00000375788.4                        1  HGNC:2665   \n",
       "...    ...                ...                      ...        ...   \n",
       "7242   NaN                NaN                      NaN        NaN   \n",
       "7243   NaN                NaN                      NaN        NaN   \n",
       "7244   NaN                NaN                      NaN        NaN   \n",
       "7245   NaN                NaN                      NaN        NaN   \n",
       "7246   NaN                NaN                      NaN        NaN   \n",
       "\n",
       "                    tag       ccdsid           havana_gene  \\\n",
       "0         [basic, CCDS]  CCDS73022.1  OTTHUMG00000036255.7   \n",
       "1         [basic, CCDS]  CCDS73022.1  OTTHUMG00000036255.7   \n",
       "2         [basic, CCDS]  CCDS73022.1  OTTHUMG00000036255.7   \n",
       "3                   NaN          NaN  OTTHUMG00000036255.7   \n",
       "4     [basic, exp_conf]          NaN  OTTHUMG00000036255.7   \n",
       "...                 ...          ...                   ...   \n",
       "7242                NaN          NaN                   NaN   \n",
       "7243                NaN          NaN                   NaN   \n",
       "7244                NaN          NaN                   NaN   \n",
       "7245                NaN          NaN                   NaN   \n",
       "7246                NaN          NaN                   NaN   \n",
       "\n",
       "         havana_transcript matchscore         seq  \n",
       "0     OTTHUMT00000088210.2        NaN         NaN  \n",
       "1     OTTHUMT00000088210.2        NaN         NaN  \n",
       "2     OTTHUMT00000088210.2        NaN         NaN  \n",
       "3                      NaN        NaN         NaN  \n",
       "4     OTTHUMT00000088209.3        NaN         NaN  \n",
       "...                    ...        ...         ...  \n",
       "7242                   NaN   5.703112  ACCAGGTGGC  \n",
       "7243                   NaN  10.495118    CCAGGTGG  \n",
       "7244                   NaN  10.495118    CCAGGTGG  \n",
       "7245                   NaN   8.212763   AGGTGGCAT  \n",
       "7246                   NaN   8.212763   AGGTGGCAT  \n",
       "\n",
       "[7247 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for col in df.columns:\n",
    "    df[col] = df[col].apply(lambda x: x[0] if isinstance(x, list) and len(x) == 1 else x)\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Use the following code to find an RE from the human genome browser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_8158/781244961.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  lst_of_id.dropna(how='all', axis=1, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320875</td>\n",
       "      <td>207320885</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.485293</td>\n",
       "      <td>GGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320875</td>\n",
       "      <td>207320885</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.485293</td>\n",
       "      <td>GGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2514</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320877</td>\n",
       "      <td>207320887</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2516</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320877</td>\n",
       "      <td>207320887</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2532</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320879</td>\n",
       "      <td>207320889</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2534</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320879</td>\n",
       "      <td>207320889</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2548</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320881</td>\n",
       "      <td>207320891</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2550</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320881</td>\n",
       "      <td>207320891</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2560</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320883</td>\n",
       "      <td>207320893</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2562</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320883</td>\n",
       "      <td>207320893</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.071769</td>\n",
       "      <td>CGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2569</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320885</td>\n",
       "      <td>207320895</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>7.795933</td>\n",
       "      <td>CGCGCGCGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2570</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320885</td>\n",
       "      <td>207320895</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>7.795933</td>\n",
       "      <td>CGCGCGCGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4223</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207321771</td>\n",
       "      <td>207321781</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>7.841279</td>\n",
       "      <td>GTCGCGCGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4224</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207321771</td>\n",
       "      <td>207321781</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>7.841279</td>\n",
       "      <td>GTCGCGCGGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid source featuretype      start        end strand              ID  \\\n",
       "2498  chr1   K562          RE  207320875  207320885      -  Tcfl5_MA0632.1   \n",
       "2500  chr1   K562          RE  207320875  207320885      +  Tcfl5_MA0632.1   \n",
       "2514  chr1   K562          RE  207320877  207320887      -  Tcfl5_MA0632.1   \n",
       "2516  chr1   K562          RE  207320877  207320887      +  Tcfl5_MA0632.1   \n",
       "2532  chr1   K562          RE  207320879  207320889      -  Tcfl5_MA0632.1   \n",
       "2534  chr1   K562          RE  207320879  207320889      +  Tcfl5_MA0632.1   \n",
       "2548  chr1   K562          RE  207320881  207320891      -  Tcfl5_MA0632.1   \n",
       "2550  chr1   K562          RE  207320881  207320891      +  Tcfl5_MA0632.1   \n",
       "2560  chr1   K562          RE  207320883  207320893      -  Tcfl5_MA0632.1   \n",
       "2562  chr1   K562          RE  207320883  207320893      +  Tcfl5_MA0632.1   \n",
       "2569  chr1   K562          RE  207320885  207320895      -  Tcfl5_MA0632.1   \n",
       "2570  chr1   K562          RE  207320885  207320895      +  Tcfl5_MA0632.1   \n",
       "4223  chr1   K562          RE  207321771  207321781      -  Tcfl5_MA0632.1   \n",
       "4224  chr1   K562          RE  207321771  207321781      +  Tcfl5_MA0632.1   \n",
       "\n",
       "      matchscore         seq  \n",
       "2498    8.485293  GGCGCGCGCG  \n",
       "2500    8.485293  GGCGCGCGCG  \n",
       "2514    8.071769  CGCGCGCGCG  \n",
       "2516    8.071769  CGCGCGCGCG  \n",
       "2532    8.071769  CGCGCGCGCG  \n",
       "2534    8.071769  CGCGCGCGCG  \n",
       "2548    8.071769  CGCGCGCGCG  \n",
       "2550    8.071769  CGCGCGCGCG  \n",
       "2560    8.071769  CGCGCGCGCG  \n",
       "2562    8.071769  CGCGCGCGCG  \n",
       "2569    7.795933  CGCGCGCGCA  \n",
       "2570    7.795933  CGCGCGCGCA  \n",
       "4223    7.841279  GTCGCGCGGC  \n",
       "4224    7.841279  GTCGCGCGGC  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_of_id = df[df[\"ID\"] == 'Tcfl5_MA0632.1'] ##Replace with the RE ID\n",
    "lst_of_id.dropna(how='all', axis=1, inplace=True)\n",
    "lst_of_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>featuretype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>matchscore</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320875</td>\n",
       "      <td>207320885</td>\n",
       "      <td>-</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.485293</td>\n",
       "      <td>GGCGCGCGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500</th>\n",
       "      <td>chr1</td>\n",
       "      <td>K562</td>\n",
       "      <td>RE</td>\n",
       "      <td>207320875</td>\n",
       "      <td>207320885</td>\n",
       "      <td>+</td>\n",
       "      <td>Tcfl5_MA0632.1</td>\n",
       "      <td>8.485293</td>\n",
       "      <td>GGCGCGCGCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid source featuretype      start        end strand              ID  \\\n",
       "2498  chr1   K562          RE  207320875  207320885      -  Tcfl5_MA0632.1   \n",
       "2500  chr1   K562          RE  207320875  207320885      +  Tcfl5_MA0632.1   \n",
       "\n",
       "      matchscore         seq  \n",
       "2498    8.485293  GGCGCGCGCG  \n",
       "2500    8.485293  GGCGCGCGCG  "
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lst_of_id[round(lst_of_id['matchscore'], 4) == 8.4853] ###Replace with matchscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['seqid', 'source', 'featuretype', 'start', 'end', 'strand', 'score',\n",
      "       'ID', 'Parent', 'gene_id', 'transcript_id', 'gene_type', 'gene_name',\n",
      "       'transcript_type', 'transcript_name', 'exon_number', 'exon_id', 'level',\n",
      "       'protein_id', 'transcript_support_level', 'hgnc_id', 'tag', 'ccdsid',\n",
      "       'havana_gene', 'havana_transcript', 'matchscore', 'seq'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "##enhancer/promoters are specific to cell type, work on how to merge data type into table\n",
    "##http://www.enhanceratlas.org/data/download/enhancer/hs/K562.bed\n",
    "##http://www.enhanceratlas.org/data/AllEPs/hs/K562_EP.txt USE THIS\n",
    "##http://www.enhanceratlas.org/Data_format_EP_v2.txt explanation\n",
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
